File: telegram-bot/schema_validation.py
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

SCHEMAS_DIR = Path(__file__).resolve().parent.parent / "schemas"


def load_schema(schema_filename: str) -> dict[str, Any]:
    schema_path = SCHEMAS_DIR / schema_filename
    with open(schema_path, "r", encoding="utf-8") as f:
        return json.load(f)


def _matches_type(value: Any, expected: str) -> bool:
    if expected == "null":
        return value is None
    if expected == "array":
        return isinstance(value, list)
    if expected == "object":
        return isinstance(value, dict)
    if expected == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if expected == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if expected == "string":
        return isinstance(value, str)
    if expected == "boolean":
        return isinstance(value, bool)
    return True


def _resolve_ref(root_schema: dict[str, Any], ref: str) -> dict[str, Any] | None:
    if not ref.startswith("#/"):
        return None

    current: Any = root_schema
    for part in ref[2:].split("/"):
        if not isinstance(current, dict) or part not in current:
            return None
        current = current[part]
    return current if isinstance(current, dict) else None


def validate_by_schema_warn_only(
    payload: Any,
    schema: dict[str, Any],
    partial: bool = False,
    path: str = "$",
    root_schema: dict[str, Any] | None = None,
) -> list[str]:
    warnings: list[str] = []
    root = root_schema or schema

    if "$ref" in schema:
        resolved = _resolve_ref(root, schema["$ref"])
        if resolved is None:
            return [f"{path}: unresolved schema ref {schema['$ref']}"]
        return validate_by_schema_warn_only(payload, resolved, partial, path, root)

    expected_types = schema.get("type")
    if expected_types is not None:
        types = expected_types if isinstance(expected_types, list) else [expected_types]
        if not any(_matches_type(payload, t) for t in types):
            actual = "array" if isinstance(payload, list) else "null" if payload is None else type(payload).__name__
            warnings.append(f"{path}: expected {'|'.join(types)} got {actual}")
            return warnings

    schema_type = schema.get("type")

    if schema_type == "object" and isinstance(payload, dict):
        properties = schema.get("properties", {})
        required = schema.get("required", [])

        if not partial and isinstance(required, list):
            for field in required:
                if field not in payload:
                    warnings.append(f"{path}.{field}: missing required field")

        for key, value in payload.items():
            if key in properties:
                warnings.extend(
                    validate_by_schema_warn_only(
                        value,
                        properties[key],
                        partial=partial,
                        path=f"{path}.{key}",
                        root_schema=root,
                    )
                )
                continue

            additional = schema.get("additionalProperties", True)
            if additional is False:
                warnings.append(f"{path}.{key}: unknown key")
                continue

            if isinstance(additional, dict):
                warnings.extend(
                    validate_by_schema_warn_only(
                        value,
                        additional,
                        partial=partial,
                        path=f"{path}.{key}",
                        root_schema=root,
                    )
                )

    if schema_type == "array" and isinstance(payload, list):
        item_schema = schema.get("items")
        if isinstance(item_schema, dict):
            for i, item in enumerate(payload):
                warnings.extend(
                    validate_by_schema_warn_only(
                        item,
                        item_schema,
                        partial=partial,
                        path=f"{path}[{i}]",
                        root_schema=root,
                    )
                )

    return warnings


def log_schema_warnings(scope: str, warnings: list[str]) -> None:
    for warning in warnings:
        logger.warning(f"[SchemaWarn:{scope}] {warning}")

File: src/utils/schemaValidation.ts
type SchemaObject = {
    $ref?: string;
    type?: string | string[];
    required?: string[];
    properties?: Record<string, SchemaObject>;
    additionalProperties?: boolean | SchemaObject;
    items?: SchemaObject;
    $defs?: Record<string, SchemaObject>;
};

interface ValidateOptions {
    partial?: boolean;
}

const isPlainObject = (value: unknown): value is Record<string, unknown> =>
    typeof value === 'object' && value !== null && !Array.isArray(value);

const matchesType = (value: unknown, expected: string) => {
    if (expected === 'null') return value === null;
    if (expected === 'array') return Array.isArray(value);
    if (expected === 'object') return isPlainObject(value);
    if (expected === 'number') return typeof value === 'number' && Number.isFinite(value);
    if (expected === 'integer') return typeof value === 'number' && Number.isInteger(value);
    return typeof value === expected;
};

const resolveRef = (root: SchemaObject, ref: string): SchemaObject | null => {
    if (!ref.startsWith('#/')) return null;
    const parts = ref.slice(2).split('/');
    let current: any = root;
    for (const part of parts) {
        if (!current || typeof current !== 'object' || !(part in current)) {
            return null;
        }
        current = current[part];
    }
    return current as SchemaObject;
};

export const validateBySchemaWarnOnly = (
    payload: unknown,
    schema: SchemaObject,
    options: ValidateOptions = {},
    path = '$',
    rootSchema: SchemaObject = schema
): string[] => {
    const warnings: string[] = [];
    const partial = options.partial === true;

    if (schema.$ref) {
        const resolved = resolveRef(rootSchema, schema.$ref);
        if (!resolved) {
            warnings.push(`${path}: unresolved schema ref ${schema.$ref}`);
            return warnings;
        }
        return validateBySchemaWarnOnly(payload, resolved, options, path, rootSchema);
    }

    const expectedTypes = schema.type ? (Array.isArray(schema.type) ? schema.type : [schema.type]) : [];
    if (expectedTypes.length > 0) {
        const typeOk = expectedTypes.some((expected) => matchesType(payload, expected));
        if (!typeOk) {
            warnings.push(`${path}: expected ${expectedTypes.join('|')} got ${Array.isArray(payload) ? 'array' : payload === null ? 'null' : typeof payload}`);
            return warnings;
        }
    }

    if (schema.type === 'object' || (Array.isArray(schema.type) && schema.type.includes('object') && isPlainObject(payload))) {
        if (!isPlainObject(payload)) return warnings;

        const props = schema.properties || {};
        if (!partial && Array.isArray(schema.required)) {
            for (const req of schema.required) {
                if (!(req in payload)) warnings.push(`${path}.${req}: missing required field`);
            }
        }

        for (const [key, value] of Object.entries(payload)) {
            if (props[key]) {
                warnings.push(...validateBySchemaWarnOnly(value, props[key], options, `${path}.${key}`, rootSchema));
                continue;
            }

            if (schema.additionalProperties === false) {
                warnings.push(`${path}.${key}: unknown key`);
                continue;
            }

            if (isPlainObject(schema.additionalProperties)) {
                warnings.push(...validateBySchemaWarnOnly(value, schema.additionalProperties, options, `${path}.${key}`, rootSchema));
            }
        }
    }

    if (schema.type === 'array' || (Array.isArray(schema.type) && schema.type.includes('array') && Array.isArray(payload))) {
        if (!Array.isArray(payload)) return warnings;
        if (schema.items) {
            payload.forEach((item, idx) => {
                warnings.push(...validateBySchemaWarnOnly(item, schema.items as SchemaObject, options, `${path}[${idx}]`, rootSchema));
            });
        }
    }

    return warnings;
};

export const logSchemaWarnings = (scope: string, warnings: string[]) => {
    warnings.forEach((warning) => {
        console.warn(`[SchemaWarn:${scope}] ${warning}`);
    });
};

